overlord/snapstate/snapmgr.go

```diff
@@ -526,6 +545,32 @@ func (m *SnapManager) ensureRefreshes() error {
 	return err
 }
 
+// ensureCatalogRefresh ensures that we refresh the catalog
+// data periodically
+func (m *SnapManager) ensureCatalogRefresh() error {
+	// sneakily don't do anything if in testing
+	if CanAutoRefresh == nil {
+		return nil
+	}
+	m.state.Lock()
```

Comment:
I need to hold the lock to get the cached store object, and to set the next timeout; I need it unlocked to call refreshCatalogs. So it'd have to be
```golang
m.state.Lock()
defer m.state.Unlock()
// do stuff
m.state.Unlock()
defer m.state.Lock()
// do other stuff
```
I thought it best to avoid the resulting spurious lock/unlock defers

Reply:
It'll be worse if we get a lock out because something in the middle panics. This would prevent even the systemctl restart that would usually happen, and potentially render a device dead.

Reply:
Ah, there's another alternative as well: splitting the logic out into a function.